Internal: Add METADATA_READ and METADATA_WRITE blocks

tlrx opened the pull request:
While looking at #3703, it looks like the usage of 'read_only' is a bit confusing.

The documentation stipulates:
> index.blocks.read_only
> Set to true to have the index read only, false to allow writes and metadata changes.

Actually, it blocks all write operations (creating, indexing or deleting a document) but also all read/write operations on index metadata. This can lead to confusion because subsequent 'indices exists' requests will return an undocumented 403 Forbidden response code where one could expect a 200 OK (as reported in #3703). Same for 'type exists' and 'alias exists' requests.

Similar issues have been reported:
- _status #2833 (deprecated in 1.2.0)
- _stats #5876
- /_cat/indices #5855

But several API calls returns a 403 when at least one index of the cluster has 'index.blocks.read_only' set to true:
```
DELETE /_all

POST /library/book/1
{
  "title": "Elasticsearch: the definitive guide"
}

POST /movies/dvd/1
{
  "title": "Elasticsearch: the movie"
}

PUT /library/_settings
{
  "index": {
    "blocks.read_only": true
  }
}
```
The following calls will return 403 / "ClusterBlockException[blocked by: [FORBIDDEN/5/index read-only (api)];]":
```
HEAD /library
HEAD /library/book
HEAD /_all/dvd

GET /library/_settings
GET /library/_mappings
GET /library/_warmers
GET /library/_aliases

GET /_settings
GET /_aliases
GET /_mappings
GET /_cat/indices
```
I agree to @imotov's comment (see https://github.com/elasticsearch/elasticsearch/pull/5876#issuecomment-42273841) and I think we should split the METADATA blocks in METADATA_READ & METADATA_WRITE blocks. This way the 'read_only' setting will block all write operations and metadata updates but the metadata will still be available.

This is a first step before updating all other actions then removing METADATA block completely.

Closes #3703
Related to #8102

tlrx commented:
@imotov @s1monw Can you have a look please? :)

tlrx commented:
For what I've seen:
`index.blocks.read_only`: if true, blocks index writes and metadata reads/writes
`index.blocks.read`: if true, blocks index reads and allows metadata reads/writes
`index.blocks.write`: if true, blocks index writes and allows metadata reads/writes

With this behavior, it's not possible to block index writes + metadata writes and allows index reads + metadata reads at the same time.

@kimchy can you have a look please? Specially regarding your comment https://github.com/elasticsearch/elasticsearch/pull/5876#issuecomment-42273841 . I find this change useful but maybe it's just me.

kimchy commented:
looking at this change, won't it be confusing to have METADATA (which is deprecated) and METADATA_READ + METADATA_WRITE while METADATA is still used across the codebase? 

I mean, I think we should either remove all usages of METADATA from the codebase, and only parse it for backward comp and resulting in 2 blocks MD_READ+MD_WRITE, and then go over all the places we use it, and use the equivalent MD_READ or MD_WRITE?

s1monw commented:
@tlrx are you picking this up again?

tlrx commented:
@s1monw yes, I asked @javanna to give me his opinion

javanna commented:
I confirm, it's on my list :)

javanna commented on rest-api-spec/test/indices.exists/20_read_only_index.yaml:
```diff
@@ -0,0 +1,30 @@
+      indices.exists:
+        index: test_index_ro
+
+  - is_true: ''
```
can you remind me what this assertion does? do we really need it here?

tlrx replied:
It's needed to check the result of the previous 'indices.exists' request.

From [here](https://github.com/elastic/elasticsearch/blob/master/src/test/java/org/elasticsearch/test/rest/client/RestResponse.java#L90):
```
 is_true: '' means the response had no body but the client returned true (caused by 200)
```

javanna replied:
alright thanks :)

javanna commented on src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -117,10 +117,10 @@ public static void registerFactory(String type, Custom.Factory factory) {
     public static final ClusterBlock INDEX_READ_BLOCK = new ClusterBlock(7, "index read (api)", false, false, RestStatus.FORBIDDEN, EnumSet.of(ClusterBlockLevel.READ));
     public static final ClusterBlock INDEX_WRITE_BLOCK = new ClusterBlock(8, "index write (api)", false, false, RestStatus.FORBIDDEN, EnumSet.of(ClusterBlockLevel.WRITE));
-    public static final ClusterBlock INDEX_METADATA_BLOCK = new ClusterBlock(9, "index metadata (api)", false, false, RestStatus.FORBIDDEN, EnumSet.of(ClusterBlockLevel.METADATA));
+    public static final ClusterBlock INDEX_METADATA_BLOCK = new ClusterBlock(9, "index metadata (api)", false, false, RestStatus.FORBIDDEN, EnumSet.of(ClusterBlockLevel.METADATA_WRITE, ClusterBlockLevel.METADATA));
```
Question: do we keep the old `ClusterBlockLevel.METADATA` as constructor argument only for bw comp right?

javanna replied:
Also, I'm confused here: `INDEX_METADATA_BLOCK` should block metadata read and metadata writes? Should we pass in `METADATA_READ` as well?

tlrx replied:
Right, there's something confusing here. METADATA must be removed and METADATA_READ must be added.

javanna replied:
sounds good

javanna commented on src/test/java/org/elasticsearch/cluster/BlockClusterStatsTests.java:
```diff
@@ -98,13 +98,6 @@ public void testBlocks() throws Exception {
-            } catch(ClusterBlockException e) {
-                assertClusterAndIndexBlocks(e);
-            }
-
```
maybe verify that exists can be called rather than deleting the whole block?

javanna commented on src/test/java/org/elasticsearch/indices/IndicesOptionsIntegrationTests.java:
```diff
@@ -801,25 +801,6 @@ public void testDeleteWarmer_wildcard() throws Exception {
     }
 
     @Test
```
why delete this test?

javanna replied:
nevermind, I see you moved it to a more specific test class, that's fine.

javanna commented on src/test/java/org/elasticsearch/indices/exists/indices/IndicesExistsTests.java:
```diff
@@ -0,0 +1,85 @@
+            try {
+                setIndexBlocks("ro", blockSetting, true);
+                assertThat(client().admin().indices().prepareExists("ro").execute().actionGet().isExists(), equalTo(true));
+            } catch (ClusterBlockException e) {
```
I think we can omit catching this and failing when caught, that's default behaviour, what matters if the `finally` I guess

javanna commented:
looks good, I left a few comments. I do agree with @kimchy on moving the whole codebase to the new blocks and maintain the old one only for bw comp, where METADATA just gets translated to METADATA_READ and METADATA_WRITE. That would isolate the bw comp aspect even more.

javanna commented on src/main/java/org/elasticsearch/cluster/block/ClusterBlockLevel.java:
```diff
@@ -29,9 +29,18 @@
 
-    public static final EnumSet<ClusterBlockLevel> ALL = EnumSet.of(READ, WRITE, METADATA);
+    /**
+     * After 1.5.0 METADATA has been splitted into two distincts cluster block levels
```
I would be more precise on the version, cause it's not clear if it is from 1.5.1 or 1.6.0.

javanna commented on src/main/java/org/elasticsearch/cluster/block/ClusterBlockLevel.java:
```diff
@@ -44,13 +53,18 @@ public int id() {
+        } else if (id == 3) {
+            return EnumSet.of(METADATA_READ);
+        } else if (id == 4) {
+            return EnumSet.of(METADATA_WRITE);
```
this `id` gets sent over the wire in `ClusterBlock#readFrom` and `ClusterBlock#writeTo`. your change makes it backwards compatible only for reads, cause a 1.6 node that gets 1.2 detects and converts it. But what happens if a newer node sends 3 or 4 to an older node? We need to add some logic based on version of nodes.

Also, I'd try and make this method package private, not sure why it's public it shouldn't IMO.

tlrx replied:
You're right. I added some logic in the ClusterBlock.writeTo() method, converting 3/4 values into 2 if the target node has a version <= 1.5.0.

tlrx replied:
This logic as been moved to ClusterBlockLevel.toId(Version version)

javanna replied:
that looks good, thanks

javanna commented on src/main/java/org/elasticsearch/action/admin/indices/mapping/delete/TransportDeleteMappingAction.java:
```diff
@@ -109,7 +109,14 @@ protected void doExecute(DeleteMappingRequest request, ActionListener<DeleteMapp
+        }
+        // DeleteMapping executes a DeleteByQuery as soon as the mapping is deleted,
+        // so we need to check the same block as DeleteByQuery here
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.WRITE, concreteIndices);
```
aren't delete mapping & delete by query gone though?

Also, how did you find this issue?

tlrx replied:
yeah... I need to rebase this branch, but not until your review is terminated :)

You can skip DeleteMappingAction and associated test.

javanna replied:
I'm afraid delete_by_query is back :)

tlrx replied:
Yes, I will add it again once I rebase (again) this PR.

> Also, how did you find this issue?

The test fails, then I dig into the code and see that DeleteMapping executes a DeleteByQuery under the hood.

tlrx replied:
Ok, DeleteByQuery is back but not DeleteMapping... so nothing to do here :)

javanna commented on src/main/java/org/elasticsearch/action/admin/indices/mapping/get/TransportGetFieldMappingsIndexAction.java:
```diff
@@ -133,6 +135,11 @@ protected GetFieldMappingsResponse newResponse() {
+    protected ClusterBlockException checkRequestBlock(ClusterState state, InternalRequest request) {
+        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_READ, request.concreteIndex());
+    }
+
```
interesting, so we were previously checking for READ (only data?), which seems like a bug, and now we only look for METADATA_READ blocks?

javanna replied:
agreed

javanna replied:
I created #10521 for this, to give this issue some more visibility. Fix is still fine in this PR, just add Closes #10521 to your commit message.

javanna commented on src/main/java/org/elasticsearch/action/admin/indices/template/put/TransportPutIndexTemplateAction.java:
```diff
@@ -64,7 +64,7 @@ protected PutIndexTemplateResponse newResponse() {
     @Override
     protected ClusterBlockException checkBlock(PutIndexTemplateRequest request, ClusterState state) {
-        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA, "");
+        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_WRITE, "");
```
not related to this change, but I get confused by these calls with empty index. Will they ever return a block? Seems like we should dig here on another PR.

tlrx replied:
I opened https://github.com/elastic/elasticsearch/issues/10530 for this.

tlrx commented:
> Also, as a follow-up might it make sense to deprecate the index.blocks.metadata setting and expose the distinction there between reads and writes now that we support it, as we do with data?

Yes, it makes sense but I'm not sure these settings will be often used. Do you think it should be added as part of this pull request or in another one?

javanna commented:
Hey @tlrx this PR is going to be big enough already :) we can open another issue for discussion when this is in. I will take care of that.

tlrx commented:
@javanna I rebased on master and updated the other APIs that were still using the deprecated METADATA block. I also integrated your previous comments. Tests are OK.

Could you please review this again when you have time? thanks :)

javanna commented on src/main/java/org/elasticsearch/action/admin/cluster/repositories/verify/TransportVerifyRepositoryAction.java:
```diff
@@ -69,7 +69,7 @@ protected VerifyRepositoryResponse newResponse() {
     @Override
     protected ClusterBlockException checkBlock(VerifyRepositoryRequest request, ClusterState state) {
-        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA, "");
+        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_WRITE, "");
```
maybe I don't know what verify repo does, bu tit sounds like a read_only operation to me. Why METADATA_WRITE and not METADATA_READ?

tlrx replied:
During repository verification proces, each node will try to write a test file into the target repository and then deleting it. That's why METADATA_WRITE seems more appropriate to me.

javanna replied:
I don't know, I believe METADATA_READ is probably enough but can't really judge, let's hear from @imotov ? ;)

tlrx replied:
Oh, I forgot to mention that the Put Repository operation (which has METADATA_WRITE) executes by default the verification operation once the repository is registered so it seems more coherent to me. Anyway I'll be glad to have Igor opinion :)

imotov replied:
Verification is called in two cases - when we create repo and when we explicitly call verification. We already check for `METADATA_WRITE` in the put repository operation, so I think we can use `METADATA_READ` here.

tlrx replied:
Ok thanks @imotov , let's change this for `METADATA_READ`.

javanna commented on src/main/java/org/elasticsearch/cluster/block/ClusterBlockLevel.java:
```diff
@@ -29,9 +30,19 @@
-    METADATA(2);
 
-    public static final EnumSet<ClusterBlockLevel> ALL = EnumSet.of(READ, WRITE, METADATA);
+    // todo-tlrx Update version number according to the branch it is merged to
```
you migth want to use //nocommit so maven notifies you when you forget about it ;)

tlrx replied:
Sure, `// todo-tlrx` is my personal Intellij marker :)

javanna commented on src/test/java/org/elasticsearch/action/admin/cluster/repositories/delete/DeleteRepositoryTests.java:
```diff
@@ -0,0 +1,55 @@
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertBlocked;
+
+@ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST)
```
why Scope.TEST? does it leave transient settings behind?

javanna replied:
I think I see why........ grrrr you remove the block but the setting stays there....... put a comment maybe?

tlrx replied:
Yes, `setClusterReadOnly(false);` update the cluster settings... I'll add a comment.

javanna commented on src/test/java/org/elasticsearch/action/admin/cluster/repositories/get/GetRepositoryTests.java:
```diff
@@ -0,0 +1,52 @@
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.hasSize;
+
+@ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST)
```
same here, does it need Scope.TEST?

tlrx replied:
Same as DeleteRepositoryTests

javanna commented on src/test/java/org/elasticsearch/action/admin/cluster/repositories/delete/DeleteRepositoryTests.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Test
+    public void testDeleteRepositoryWithBlocks() {
+        Client client = client();
```
is it important to use the same client for every call? why not calling client() at every request?

tlrx replied:
Not important, I remove this in favor of calling client().

javanna commented on src/test/java/org/elasticsearch/action/admin/cluster/repositories/verify/VerifyRepositoryTests.java:
```diff
@@ -0,0 +1,57 @@
+        VerifyRepositoryResponse response = client.admin().cluster().prepareVerifyRepository("test-repo-blocks").execute().actionGet();
+        assertThat(response.getNodes().length, equalTo(cluster().numDataAndMasterNodes()));
+    }
+}
```
I wonder if we should merge all the above repository tests in `RepositoryTests` or something along those lines

tlrx replied:
Right, I merge the repository tests in `RepositoryBlocksTests` class.

javanna commented on src/test/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotStatusTests.java:
```diff
@@ -0,0 +1,44 @@
+            setClusterReadOnly(false);
+        }
+    }
+}
```
here too I wonder if we should merge all these snashot tests in the same class, since each one of them have a single test method.

tlrx replied:
Same here, I merge the repository tests in a SnapshotBlocksTests class.

javanna commented on src/test/java/org/elasticsearch/indices/template/SimpleIndexTemplateTests.java:
```diff
@@ -54,6 +56,7 @@
 /**
  *
  */
+@ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST)
```
I think this new test should be added to a separate new class, otherwise all of the existing methods will need to recreate the cluster before their execution, it's going to slow things down I'm afraid

tlrx replied:
Right, I should have thought of that... I'll add a IndexTemplateBlocksTests class.

javanna commented on src/test/java/org/elasticsearch/indices/warmer/SimpleIndicesWarmerTests.java:
```diff
@@ -43,14 +43,20 @@
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertBlocked;
 import static org.hamcrest.Matchers.*;
 
+@ClusterScope(scope = ElasticsearchIntegrationTest.Scope.TEST)
```
same as above

tlrx replied:
Ok, I'll add IndicesWarmerBlocksTests class.

javanna commented on src/test/java/org/elasticsearch/test/ElasticsearchIntegrationTest.java:
```diff
@@ -1418,6 +1418,34 @@ public static void enableTranslogFlush(String index) {
+    }
+
+    /** Disables an index block for the specified index using a given {@link Client} */
+    public static void disableIndexBlock(Client client, String index, String block) {
```
is it important to provide a specific client here? isn't the previous method enough?

tlrx replied:
I added this method so that classes which extends `ElasticsearchSingleNodeTest` were also able to use this method.

Since it was the case in only 2 classes, I removed the method with the specific client and created two classes which extends `ElasticsearchIntegrationTest`: `IndicesStatsBlocksTests` and `IndicesSegmentsBlocksTests`.

javanna commented on src/test/java/org/elasticsearch/test/ElasticsearchIntegrationTest.java:
```diff
@@ -1418,6 +1418,34 @@ public static void enableTranslogFlush(String index) {
+    }
+
+    /** Enables an index block for the specified index using a given {@link Client} */
+    public static void enableIndexBlock(Client client, String index, String block) { // todo-tlrx voir comment améliorer ça
```
same as above, can we get rid of this variant?

javanna commented on src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java:
```diff
@@ -117,6 +120,31 @@ public static void assertAcked(DeleteIndexResponse response) {
         assertVersionSerializable(response);
     }
 
+    public static void assertBlocked(ActionRequestBuilder builder) {
```
may I ask you to add javadocs here for people using our test-jar?

javanna commented on src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java:
```diff
@@ -117,6 +120,31 @@ public static void assertAcked(DeleteIndexResponse response) {
+        assertBlocked(builder, null);
+    }
+
+    public static void assertBlocked(ActionRequestBuilder builder, ClusterBlock expectedBlock) {
```
javadocs here too?

javanna commented:
left a bunch of minor comments @tlrx looks great though

tlrx commented:
@javanna thanks again for your review! I just updated this pr with your last comments.

javanna commented on src/main/java/org/elasticsearch/action/admin/indices/warmer/put/TransportPutWarmerAction.java:
```diff
@@ -80,7 +80,13 @@ protected PutWarmerResponse newResponse() {
+        }
+        // PutWarmer executes a SearchQuery before adding the new warmer to the cluster state,
+        // so we need to check the same block as TransportSearchTypeAction here
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.READ, concreteIndices);
```
Opened #10522 , please reference this one too in your commit message ;)

tlrx replied:
Ok

javanna commented on src/main/java/org/elasticsearch/cluster/block/ClusterBlockLevel.java:
```diff
@@ -44,13 +55,35 @@ public int id() {
+    public int toId(Version version) {
+        //nocommit Update version number according to the branch it is merged to
+        if ((version != null) && (version.onOrBefore(Version.V_1_5_0))) {
+            if ((this == ClusterBlockLevel.METADATA_READ || this == ClusterBlockLevel.METADATA_WRITE)) {
```
you can remove one level of parentheses right?

tlrx replied:
Sure!

javanna commented on src/main/java/org/elasticsearch/cluster/block/ClusterBlockLevel.java:
```diff
@@ -44,13 +55,35 @@ public int id() {
+     */
+    public int toId(Version version) {
+        //nocommit Update version number according to the branch it is merged to
+        if ((version != null) && (version.onOrBefore(Version.V_1_5_0))) {
```
the version null check is a bit scary, when can the version be null? I would rather add an assert, so we catch it if it happens

tlrx replied:
Ok

javanna commented on src/test/java/org/elasticsearch/cluster/BlockClusterStatsTests.java:
```diff
@@ -41,7 +44,7 @@
 
```
I have the feeling that this test became obsolete after all the specific tests that you added. Maybe we can remove this class entirely?

tlrx replied:
Yes, everything is covered in other test classes. I also think we can removed it now.

javanna commented on src/test/java/org/elasticsearch/indices/template/IndexTemplateBlocksTests.java:
```diff
@@ -0,0 +1,76 @@
+    @Test
+    public void testIndexTemplatesWithBlocks() throws IOException {
+        // clean all templates setup by the framework.
+        client().admin().indices().prepareDeleteTemplate("*").get();
```
out of curiosity, why is this necessary?

tlrx replied:
ElasticsearchIntegrationTest creates some random index template before each test (see ElasticsearchIntegrationTest.randomIndexTemplate()). I did not find a way to disable this.

javanna replied:
I get that, I was just wondering why those default templates bother here

tlrx replied:
You're right it doesn't bother, the test can be simplified. Thanks for spotting it.

javanna commented on src/test/java/org/elasticsearch/indices/warmer/IndicesWarmerBlocksTests.java:
```diff
@@ -0,0 +1,160 @@
+        createIndex("test-blocks");
+        ensureGreen("test-blocks");
+
+        int warmerId = 0;
```
not sure why you need the warmer name to change all the time, you can just keep submitting the same name no?

tlrx replied:
In this test, there is one case the warmer is not blocked but acked so it can't be submitted again... but you're right, I'll simplify this :)

javanna commented:
looks great @tlrx can you rebase so I have a last look please?

tlrx commented:
@javanna Done!

javanna commented:
Just one last nitpick: shall we use the same naming convention for the new test classes that you introduced? e.g. FooBlocksTests . When I look at DeleteIndexTests and PendingTasksTests for instance, i wouldn't want other generic tests to be added to the same class in the future, given that the scope is TEST, I thought naming might help telling what the test is about.

Beyond this, LGTM! @imotov do you want to have a final look at it?

javanna commented on src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/TransportCreateSnapshotAction.java:
```diff
@@ -65,7 +65,7 @@ protected CreateSnapshotResponse newResponse() {
     @Override
     protected ClusterBlockException checkBlock(CreateSnapshotRequest request, ClusterState state) {
-        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA, "");
+        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_WRITE, "");
```
maybe we should only check for metadata_read here? then #8102 would get fixed too. create snapshot only writes to the repo right? it's not really a write operation within the cluster? As long as you can read the data and metadata, you should be able to snapshot no? which makes me wonder, do we check for index read blocks on data?

@imotov thoughts?

imotov replied:
This is tough one. During snapshot operation, we create a record about running snapshot in metadata. So, technically we write into metadata. It's not a persistent record though.... so it all depends on the definition of the metadata. Maybe it was a mistake to put snapshot into metadata though. Now I am thinking it should have been custom cluster state level element. So, we might want to move snapshot into cluster state custom and use READ here.

javanna replied:
I see why this is is confusing. But if we move this to METADATA_READ, will we be able to make a snapshot when the cluster is in read_only mode? If so, I would consider changing to METADATA_READ for now, which fixes #8102 too , add a comment that explains why (pretty much Igor's comment above) and next step could be to move snapshot out of metadata even?

javanna replied:
Bottom line: I am fine with not addressing #8102 on this PR, we can have another one afterwards, I would just love to have that issue fixed soon too, regardless of whether we move snapshots to customs (which seems a bigger change).

javanna commented on src/main/java/org/elasticsearch/action/admin/cluster/snapshots/delete/TransportDeleteSnapshotAction.java:
```diff
@@ -64,7 +64,7 @@ protected DeleteSnapshotResponse newResponse() {
     @Override
     protected ClusterBlockException checkBlock(DeleteSnapshotRequest request, ClusterState state) {
-        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA, "");
+        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_WRITE, "");
```
same as above, I guess this should be consistent with what we chose above. Do we even have to check for blocks here though?

javanna replied:
@imotov what do you think here? METADATA_READ or METADATA_WRITE? is it the same problem as described above?

imotov replied:
Let's stick with `METADATA_WRITE`.

javanna replied:
can you explain why? :)

imotov replied:
.... because it remove a record created in create snapshot. So, if it's METADATA_WRITE there it should be METADATA_WRITE here.

javanna commented on src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/TransportRestoreSnapshotAction.java:
```diff
@@ -65,7 +65,7 @@ protected RestoreSnapshotResponse newResponse() {
     @Override
     protected ClusterBlockException checkBlock(RestoreSnapshotRequest request, ClusterState state) {
-        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA, "");
+        return state.blocks().indexBlockedException(ClusterBlockLevel.METADATA_WRITE, "");
```
one more doubt, shouldn't we also check for blocks on data here?

javanna replied:
@imotov thoughts here? or the fact that we know the index must be closed makes it already good to go?

imotov replied:
This might be `WRITE` actually - we are creating an index or modifying an index after all.

javanna replied:
but we might also restore metadata no? I think it requires both?

imotov replied:
Indeed we might restore persistent settings and templates. We need both.

imotov commented on src/main/java/org/elasticsearch/action/admin/indices/optimize/TransportOptimizeAction.java:
```diff
@@ -121,11 +121,11 @@ protected GroupShardsIterator shards(ClusterState clusterState, OptimizeRequest
     @Override
     protected ClusterBlockException checkGlobalBlock(ClusterState state, OptimizeRequest request) {
-        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA);
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
```
I am always getting confused by this one and refresh. Shouldn't this be `WRITE` and not `METADATA_WRITE`? We don't really change any metadata here.

rjernst replied:
I don't think optimize should be `WRITE`.  It is an administrative type action, you aren't actually passing data, just telling the system to do something.

imotov replied:
`METADATA_READ` then?

tlrx replied:
When I updated this block I supposed that users want to block refresh/optimize actions on a read only cluster so I set `METADATA_WRITE`. Anyway I don't have a strong opinion on this and `METADATA_READ` is fine to me since it will still be blockable with the `index.blocks.metadata` setting.

rjernst replied:
IMO it is write because the cluster is changing (an action is happening in the background), and metadata because it is not changing data (it is changing the underlying storage, but from the user perspective the data is not changing).

imotov replied:
I don't have strong feelings about it. I just find it confusing.

tlrx replied:
Can we agree on letting `METADATA_WRITE` ?

imotov replied:
+1 on keeping it `METADATA_WRITE`

javanna replied:
I agree this is confusing, given how we classified operations in this PR, I think this one is more under metadata_write than metadata_read. We might need to adress this confusion at some point though (same with snapshots). So +1 on leaving METADATA_WRITE.

imotov commented:
Left a couple of comments. Otherwise LGTM.

tlrx commented:
Thanks @imotov and @javanna for your reviews. I made few changes according to your last comments.

Concerning #8102 I think it should be the subject of a dedicated pull request... I'd like to know more how restoring an index works when blocks are enabled and moving the cluster state from metadata to custom is not so simple I guess? Since this pull request doesn't change the current behavior I think it could be merged as it.

javanna commented on src/main/java/org/elasticsearch/cluster/block/ClusterBlock.java:
```diff
@@ -131,7 +131,7 @@ public void readFrom(StreamInput in) throws IOException {
         final int len = in.readVInt();
         ArrayList<ClusterBlockLevel> levels = new ArrayList<>();
         for (int i = 0; i < len; i++) {
```
Given that we changed the serialization code for this class, would it be possible to write a small unit test that verifies it?  We already have similar `*RequestTests#testSerialization` tests that randomize the version and make sure that everything is fine.

imotov commented:
@tlrx  Sorry, missed the restore one the last time around. Otherwise, I think it's good.

tlrx commented:
@javanna I just added the test, can you please have a look? @imotov no worries, thanks for your review.

I added another check on `WRITE` block on the empty `""` index in the `TransportRestoreSnapshotAction`. I think that there's no need to check for `WRITE` block for the indices specified in the request as @javanna suggested in [this comment](https://github.com/elastic/elasticsearch/pull/9203#discussion_r28140550) because the indices are necessarily in closed state and so `WRITE` are *de facto* blocked.

javanna commented on src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java:
```diff
@@ -0,0 +1,85 @@
+            Version fromVersion = randomVersion();
+            Version toVersion = randomVersion();
+
+            ClusterBlockLevel[] blockLevels = ALL_CLUSTER_BLOCK_LEVELS;
```
`ClusterBlockLevel.values()` ? don't think we need the constant

javanna commented on src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java:
```diff
@@ -0,0 +1,85 @@
+            assertThat(result.retryable(), equalTo(clusterBlock.retryable()));
+            assertThat(result.disableStatePersistence(), equalTo(clusterBlock.disableStatePersistence()));
+
+            for (ClusterBlockLevel level : clusterBlock.levels()) {
```
can you also check that result.levels.size() == clusterBlock.levels.size() ?

tlrx replied:
I think we cannot check this equality because `METADATA` is deserialized as `METADATA_WRITE + METADATA_READ`.

javanna replied:
right, good point...I wish there was a way though to make sure that the result doesn't contain more stuff that we don't go over, as we iterate on clusterBlock only. You see what I mean?

tlrx replied:
Sure. I added [the following assertion](https://github.com/elastic/elasticsearch/blob/master/src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java#L81) before merging this pr.

javanna replied:
fantastic thanks a lot

javanna commented on src/test/java/org/elasticsearch/cluster/block/ClusterBlockTests.java:
```diff
@@ -0,0 +1,85 @@
+
+            ClusterBlockLevel[] blockLevels = ALL_CLUSTER_BLOCK_LEVELS;
+            if (fromVersion.before(Version.V_1_6_0)) {
+                blockLevels = new ClusterBlockLevel[]{READ, WRITE, METADATA};
```
I find this block a bit confusing, as well as `fromVersion` because it doesn't get used to serialize. I do see why you are doing this, but maybe it is not needed, we can just randomize the blocks (deprecated one will get mixed up with new ones, no big deal). What do you think? And rename `toVersion` to just `version`?

tlrx replied:
Ok

javanna commented:
The new serialization test looks good, left a couple of minor comments. When you backport to 1.x, can you double check that our existing bw comp tests already cover some potential bw comp problems here (e.g. ClusterBlocks gets serialized and everything is fine)? We didn't write a specific bw comp tests but only a unit test for serialization, I think this is enough but I didn't verify it.

Other than that LGTM.

javanna commented:
LGTM @tlrx thanks!!!

tlrx commented:
This pr has been merged in master and 1.x. For this latest branch I added *BlocksTests for the IndicesStatus and DeleteMapping operations which are removed in master branch.

There's a bwc test for cluster state updates but it doesn't use any blocks. I'll add it one in a dedicated pull request.

Thanks again @javanna and @imotov  for your very helpful reviews.
